numba/misc/coverage_support.py

```diff
@@ -0,0 +1,95 @@
+"""
+Implement code coverage support.
+
+Currently contains logic to extend ``coverage`` with line covered by the
+compiler.
+"""
+from typing import Optional, Sequence, Callable, no_type_check
+from collections import defaultdict
+from abc import ABC, abstractmethod
+import atexit
+from functools import cache
+
+from numba.core import ir
+
+
+try:
+    import coverage
+except ImportError:
+    coverage_available = False
+else:
+    coverage_available = True
+
+
+@no_type_check
+def get_active_coverage():
+    """Get active coverage instance or return None if not found.
+    """
+    cov = None
+    if coverage_available:
+        cov = coverage.Coverage.current()
+    return cov
+
+
+_the_registry: Callable[[], Optional["NotifyLocBase"]] = []
+
+
+def get_registered_loc_notify() -> Sequence["NotifyLocBase"]:
+    return list(filter(lambda x: x is not None,
+                       (factory() for factory in _the_registry)))
+
+
+@cache
+def _get_coverage_data():
+    """
+    Make a singleton ``CoverageData``.
+    Avoid writing to disk. Other processes can corrupt the file.
+    """
+    covdata = coverage.CoverageData(no_disk=True)
+    cov = get_active_coverage()
+    assert cov is not None, "no active Coverage instance"
+
+    @atexit.register
+    def _finalize():
+        cov.get_data().update(covdata)
```

Comment:
Note to self. This is the part that updates the existing active coverage with the information from the additional custom coverage object.